src/storage/transaction/ChainAddEdgesLocalProcessor.cpp

```diff
@@ -363,9 +288,31 @@ cpp2::AddEdgesRequest ChainAddEdgesLocalProcessor::reverseRequest(
 }
 
 void ChainAddEdgesLocalProcessor::finish() {
-  VLOG(1) << uuid_ << " commitLocal(), code_ = " << apache::thrift::util::enumNameSafe(code_);
-  pushResultCode(code_, localPartId_);
-  finished_.setValue(code_);
+  auto rc = (rcPrepare_ == Code::SUCCEEDED) ? rcCommit_ : rcPrepare_;
+  if (rcPrepare_ == Code::SUCCEEDED) {
+    VLOG(1) << uuid_ << execDesc_ << makeReadableEdge(req_)
+            << ", rcPrepare_=" << apache::thrift::util::enumNameSafe(rcPrepare_)
+            << ", rcRemote_=" << apache::thrift::util::enumNameSafe(rcRemote_)
+            << ", rcCommit_=" << apache::thrift::util::enumNameSafe(rcCommit_);
+  }
+  do {
+    if (rcPrepare_ != Code::SUCCEEDED) {
+      break;  // nothing written, no need to recover.
+    }
+
+    if (rcCommit_ != Code::SUCCEEDED) {
+      reportFailed(ResumeType::RESUME_CHAIN);
```

Comment:
I see `addPrime` is called in `reportFailed`, why we need to do it again?

Reply:
In my opinion,  "addPrime()" is really hard to understand.
"reportFailed" is just a self explain/comment function .